File: typescript/lruCache.ts
type Node<T> = {
    value: T,
    next?:Node<T>,
    prev?:Node<T>
}

export default class LRU<K,V> {
    private length:number;
    private head?: Node<V>;
    private tail?: Node<V>;
    private capacity: number

    private lookup: Map<K,Node<V>>;
    private reverseLookup: Map<Node<V>,K>

    constructor(capacity = 10){
        this.length = 0
        this.head = undefined
        this.tail = undefined
        this.capacity = capacity
        this.lookup = new Map<K,Node<V>>()
        this.reverseLookup = new Map<Node<V>,K>()

    }
    private createNode(value: V){
        const node = { value } as Node<V>
        return node
    }

    update(key: K ,value: V){
        let node = this.lookup.get(key)
        if(!node){
            node = this.createNode(value)
            this.length++
            this.prepend(node)
            this.trimCache()

            this.lookup.set(key,node)
            this.reverseLookup.set(node,key)
            
        }else{
            this.detach(node)
            this.prepend(node)
            node.value = value 
        }
    }

    get(key: K){
        const node = this.lookup.get(key)
        if(!node){
            return undefined
        }
        this.detach(node)
        this.prepend(node)
        return node?.value
    }

    private detach(node: Node<V>){
        const prev_node: Node<V> | undefined = node.prev
        const next_node: Node<V> | undefined = node.next
        if(this.head === node){
            this.head = this.head.next
        }
        if(this.tail === node){
            this.tail = this.tail.prev
        }
        
        if(prev_node) prev_node.next = next_node
        if(next_node) next_node.prev = prev_node

        node.prev = undefined
        node.next = undefined 


    }
    private prepend(node: Node<V>){
        if(!this.head){
            this.head = this.tail = node
            return
        }
        node.next = this.head
        this.head.prev = node
        this.head = node
    }
    private trimCache(): void {
        if(this.length <= this.capacity){
            return
        }

        const tail = this.tail as Node<V>
        this.detach(tail)
        this.length --

        const key = this.reverseLookup.get(tail)
        this.lookup.delete(key)
        this.reverseLookup.delete(tail)
    }


}

File: golang/ds/lruCache.go
package ds

import "fmt"

type CacheNode struct {
	val  int
	next *CacheNode
	prev *CacheNode
}

type LruCache struct {
	Head          *CacheNode
	Tail          *CacheNode
	Capacity      int
	Length        int
	Lookup        map[int]*CacheNode
	ReverseLookup map[*CacheNode]int
}

type Lru interface {
	Update()
	Get() int
}

func (l *LruCache) Update(key int, val int) {
	if cacheNode, found := l.Lookup[key]; found {
		l.Detach(cacheNode)
		cacheNode.val = val
		l.Prepend(cacheNode)
	} else {
		node := CreateNode(val)
		l.Prepend(node)
		l.Length++
		l.ArrangeCache()

		l.Lookup[key] = l.Head
		l.ReverseLookup[l.Head] = key
	}
}

func CreateNode(val int) *CacheNode {
	return &CacheNode{
		val: val,
	}
}

func (l *LruCache) Detach(node *CacheNode) {

	if l.Head == node {
		l.Head = l.Head.next
	}
	if l.Tail == node {
		l.Tail = l.Tail.prev
	}

	if node.prev != nil && node.next != nil {
		node.next.prev = node.prev
		node.prev.next = node.next
	}

	node.next = nil
	node.prev = nil
}

func (l *LruCache) Prepend(node *CacheNode) {
	if l.Head == nil {
		l.Head = node
		l.Tail = l.Head
		return
	}
	node.next = l.Head
	l.Head.prev = node
	l.Head = node

}

func (l *LruCache) ArrangeCache() {
	if l.Length <= l.Capacity {
		return
	}
	tmp := l.Tail
	l.Detach(tmp)
	tailKey := l.ReverseLookup[tmp]
	delete(l.Lookup, tailKey)
	delete(l.ReverseLookup, tmp)
	l.Length--

}

func (l *LruCache) Get(key int) int {
	if node, found := l.Lookup[key]; found {
		return node.val
	} else {
		return -1
	}
}
func (l *LruCache) DisplayList() {
	curr := l.Head
	for curr != l.Tail.next {
		fmt.Println(curr.val)
		curr = curr.next
	}
}
